### DISCUSSION TITLE
GSoC 2024: Extended functionality

### DISCUSSION THREAD

**Discussion by GuChad369** [pgmoneta/pgmoneta#287] [Ideas]
## Intro

Hi,

My name is Chao Gu, and I have been selected as a contributor for the `GSoC 2024` program with the pgmoneta project. I will be working on implementing extended functionality for pgmoneta.

Currently, pgmoneta is focused on implementing delta backup. To achieve this, pgmoneta requires direct access to files on the PostgreSQL server side. My project aims to lay the foundation for this delta backup functionality.

My main goal for this extended functionality project is as follows:
- Develop a basic version of `pgmoneta_ext` that accomplishes specific functions using add-on modules PostgreSQL extension.
- Further enhance the `pgmoneta_ext` function to retrieve metric information regarding backup and restore processes.
- Implement additional permissions to access more files from the server instance.
- Employ efficient strategies for compression and backup to support the delta backup infrastructure.
- Provide comprehensive documentation for the `pgmoneta_ext` function, detailing its structure, usage, and troubleshooting guidelines.

For more details, please review my [proposal](https://drive.google.com/file/d/1Qpv13V4snmkZ__Lw7kByg3ACxQGGIT_B/view?usp=sharing). 

For more precise updates, please look at the progress log below.

I am open to new suggestions and ideas, feel free to either discuss it with me here or through email. You can contact me at chadraven369@gmail.com.

We've created the repository for [pgmoneta_ext](https://github.com/pgmoneta/pgmoneta_ext). You can access the source code via the provided link.

## List of the work done:

Midterm:
  - [x] Initiate the basic skeleton of `pgmoneta_ext` based on the `PostgreSQL` extension structure.
      - [x] Implement functionality to return the project version number in a `Datum` [[#2] Initial pgmoneta_ext skeleton #1](https://github.com/pgmoneta/pgmoneta_ext/pull/1)
      - [x] Create basic documentation for the user and developer manual [[#3] Initial documentation #6](https://github.com/pgmoneta/pgmoneta_ext/pull/6)
  - [x] Add `pg_switch_wal` and `CHECKPOINT` implementation to the extension.
      - [x]  Integrate `pg_switch_wal` to trigger WAL switch operation [[#4] SQL: pg_switch_wal() #7](https://github.com/pgmoneta/pgmoneta_ext/pull/7)
      - [x] Implement `CHECKPOINT` to force `PostgreSQL` to carry out an immediate checkpoint [[#5] SQL: CHECKPOINT #8](https://github.com/pgmoneta/pgmoneta_ext/pull/8)
  - [x] Create a test suite to test extension functions [[#9] Initial test suite #11](https://github.com/pgmoneta/pgmoneta_ext/pull/11)
  - [x] Two small tasks completed.
    - [x] Enabled RPM support for the project. Refer to [[#12] RPM support #14](https://github.com/pgmoneta/pgmoneta_ext/pull/14).
    - [x] Created a new function to return the role of the current user. Refer to [[#10] pgmoneta_ext_check_privilege #13](https://github.com/pgmoneta/pgmoneta_ext/pull/13).
  - [x] Integrate extension
    - [x] Add a test suite for `pgmoneta` in Docker container #321. 
    - [x] Create a `extension.h|.c` to call extension functions #336.
  - [x] Add new CLI commands `compress`/`decompress`#349.
  - [x] Get the database OID through the extension on the server side [[#18] Get the specific database OID by using the database name #19](https://github.com/pgmoneta/pgmoneta_ext/pull/19)
  - [x] Create two functions to retrieve `extra` files in PostgreSQL  [[#15] Get extra files #17](https://github.com/pgmoneta/pgmoneta_ext/pull/17)
  - [x] Retrieve extra files on the server side when the user sets the `extra` parameter #341 
  - [x] Follow the core project manual to update the current manual [[#20] Manuals for 0.1.0 #21](https://github.com/pgmoneta/pgmoneta_ext/pull/21)

Final:
 - [x] Initiated the function `pgmoneta_ext_full_backup` and store the necessary data to set up the structure[[#22] Initial pgmoneta_ext_full_backup() #24](https://github.com/pgmoneta/pgmoneta_ext/pull/24)
 - [x] Created three functions to receive the file from the client side, which will help us obtain the manifest file from the client. This allows the extension to compare the manifest to perform incremental backups[[#22] Receive file from client #25](https://github.com/pgmoneta/pgmoneta_ext/pull/25)
 - [x] The core-side implementation for sending the manifest file #387
 - [x] Restrict user input to avoid unknown types for the CLI option `-F` #390
 - [x] Improve compatibility for base64 encode and decode functions #393
 - [x] Implement four compression types to enable CLI option support for compressing wire protocol #391
 - [x] Implement three aes mode to enable CLI option support for encryption wire protocol #406

Other related pr:
 - [x] #361 
 - [x] [[#22] Get LSN #23](https://github.com/pgmoneta/pgmoneta_ext/pull/23)
 
Future jobs:
 - [ ] Compare the files with the server's PostgreSQL instance.
 - [ ] Handle the result.

Please refer to the issue in `pgmoneta_ext` on [Incremental Backup](https://github.com/pgmoneta/pgmoneta_ext/pull/25).

### 10/26/2024
Finally, we’ve implemented the receive file chunk functionality in `pgmoneta_ext`. For more details, please refer to the code in the [PR for `pgmoneta_ext`](https://github.com/pgmoneta/pgmoneta_ext/pull/25) and the related send file function in the core project, available in the [PR for `pgmoneta`](https://github.com/pgmoneta/pgmoneta/pull/387).

With this, our basic infrastructure is in place. The remaining task is to implement the compare function in `pgmoneta_ext`. Huge thanks to @jesperpedersen  and @Jubilee101 !

### 10/19/2024
This week, we primarily focused on adding CLI support for wire protocol encryption. I implemented three encryption modes that handle encrypting and decrypting strings into buffers and integrated them into the process.  For more details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/406) and the related [issue](https://github.com/pgmoneta/pgmoneta/issues/377).

### 10/13/2024
This week, we primarily focused on two tasks:

- Updating the base64 encode and decode function signatures to improve compatibility with various data types. For more details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/393) and the related [issue](https://github.com/pgmoneta/pgmoneta/issues/392).
  
- Continuing to enhance CLI support for the wire protocol. I implemented four types of compression functions that compress strings into buffers and integrated them into the process. For further details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/391) and the related [issue](https://github.com/pgmoneta/pgmoneta/issues/376).

### 10/06/2024
This week, we mainly focused on the following three tasks:

- Set up the basic structure for sending files from the client to the server in the core project. I will continue to develop and test all the functions. For further details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/387) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/147).
  
- Fixed a small bug in the CLI option `-F` to restrict user input and avoid unknown types. For further details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/390) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/389).
  
- Worked on adding CLI support for JSON compression. I have already implemented the gzip functions, but there are three more compression methods to complete. For further details, please refer to the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/376).

### 09/28/2024
This week, I experimented with several methods to send the file from the client to the extension side and initiated some code, but none of them worked very well. However, I found one approach that might work, similar to our previous function, `pgmoneta_ext_get_file`, which sends a file from the extension to the client. I have implemented the extension-side code, which uses three functions to handle the process. Once this test is successful, I will begin implementing the core side, which will simply call these functions using our current core protocol to execute the command. The core side is not complex. For further details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/25) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/22).

### 09/22/2024
This week, we discussed the [Streaming Replication Protocol](https://www.postgresql.org/docs/17/protocol-replication.html) and [Message Formats](https://www.postgresql.org/docs/devel/protocol-message-formats.html). I’ve reviewed the relevant code in our core project, and next week, I will work on setting up the initial version of the code to send a file from the core to the extension side.

### 09/15/2024
This week, we discussed the functions `pg_backup_start()` and `pg_backup_stop()`. Following this, we initiated the function `pgmoneta_ext_full_backup` to call these two functions and store the necessary data to set up the structure for the backup. For further details, please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/24) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/22).

### 09/07/2024
This week, we worked on implementing the retrieval of the LSN using the internal function `GetXLogWriteRecPtr` and tested it across all five PostgreSQL versions. It worked as expected. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/23) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/22) for further details.

### 09/01/2024
This week, we discussed how to implement our goal of incremental backup and how to retrieve the start/end LSN. We found a function, `GetXLogWriteRecPtr`, that can help us get the current write pointer of the WAL.

### 08/23/2024
This week, we focused on researching the [`pg_dump`](https://doxygen.postgresql.org/pg__dump_8c_source.html) source code, particularly the `-Fc` option, to investigate how `pg_dump` achieves specific database backups and to explore how we might simulate this. The conclusion is that this would require a significant amount of work and might not be suitable for our extension project at this time. We will continue to develop new APIs for the extension, and there may be other ways to achieve this goal.

We have also started working on implementing **incremental backups** using the extension for our core project. The initial approach to sending the manifest file to the extension involved creating a new table to write the file. This is detailed in [PR](https://github.com/pgmoneta/pgmoneta/pull/361). However, this approach may not be ideal, as users might not be satisfied with it. We need to consider other methods, especially by researching [`pgbackrest`](https://pgbackrest.org/user-guide-rhel.html), to find a better solution.

### 08/14/2024
Before the midterm evaluation, this week we are preparing our `pgmoneta_ext` manual for the 0.1.0 release and completing the test suite to include the new functions. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/21) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/20) for further details.

### 08/09/2024
This week, we focused on improving the process of retrieving extra files and fixing bugs in both the core and extension projects.

For the extension project, we added a new test suite for the new functions and ensured compatibility with PostgreSQL version 17. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/17) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/15) for further details.

For the core project, we introduced a new `info` feature that allows information about `extra` to be displayed when the user calls the `info` command. We also fixed some bugs. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/341) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/328) for further details.

### 08/02/2024
I have been researching `pg_dump` for PostgreSQL and reading some of its source code to understand its workflow. Additionally, I have been learning about the `BASE_BACKUP` protocol in the core project to identify the differences between `pg_dump` and `BASE_BACKUP`. Our goal is to find a way to simulate `pg_dump` to achieve specific database backups.

Currently, our idea is to first find the specific database OID using an extension. I have added new functions, `pgmoneta_ext_get_oid` and `pgmoneta_ext_get_oids`, to return the specific database OID and all database OIDs, respectively. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/19) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/18) for further details.

The next step will be to learn how `pg_dump` executes the backup and try to simulate it using the extension to achieve our goal.

### 07/30/2024
I spent several days adding CLI commands `compress` and `decompress` to allow users to perform compression using the configured method. This is similar to `encrypt` and `decrypt`; the code is the same, but the execution function is changed to handle compression.

Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/349) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/345) for further details.

### 07/26/2024
Create a new message, `pgmoneta_receive_extra_files`, to receive extra files. This will call the extension function to get files from the server. Additionally, we will add a new workflow, `wa_extra.c`, which will be integrated into the backup workflow. When the user initiates a backup, it automatically retrieves extra files from the source specified in the configuration. Users can set the desired files in the configuration using `extra = path/to/file1, path/to/dir`.

Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/341) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/328) for further details.

### 07/19/2024
Add two new extension functions in `pgmoneta_ext` to assist the core project `pgmoneta` in retrieving files from the server side:

- `pgmoneta_ext_get_file`: This function can retrieve a specified file. Currently, it is limited to superusers and does not restrict access to any directories.
- `pgmoneta_ext_get_files`: This function can retrieve all file paths in a specified directory and return them in an array. Currently, it is limited to superusers.

Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/17) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/15) for further details.

### 07/12/2024
Add a test suite for `pgmoneta`. This process took weeks to complete. Initially, we used a script to run the tests, which could affect the existing `pgmoneta`. After discussions, we moved the test into a container that installs `PostgreSQL`, `pgmoneta`, and `pgmoneta_ext` within the container. This approach ensures that nothing on the user's current machine is affected.

Although it works well, the problem is that the runtime is very long. We continue improving the installation process to make it quicker for users to test. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/321) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/319) for further details.

For the `extension.h|.c`, which enables `pgmoneta` to query the extension functions using the wire protocol, we have just the basic structure for the extension. As the extension grows, its functionality will be improved. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta/pull/336) and the associated [issue](https://github.com/pgmoneta/pgmoneta/issues/319) for further details.

### 06/25/2024
Complete the following tasks:

- Add a `pgmoneta_ext.spec` file and the associated documentation to enable RPM support for the project. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/14) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/12) for further details.

- Create a new function `pgmoneta_ext_check_privilege` that returns the role of the current user. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/13) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/10) for further details.

### 06/22/2024
We got the [[#9] Initial test suite #11](https://github.com/pgmoneta/pgmoneta_ext/pull/11) merged. This is really important for our project.

The test suite makes downloading PostgreSQL easier and ensures the specific version is installed. We mainly use the `REL_XX_STABLE` branch from the PostgreSQL mirror repository, which is very stable. We can continue to add new versions to our script according to the official repository.

It also helps make our installation and testing more convenient. I learned a lot from this test suite. Thank you for any help during this process; I appreciate it.

### 06/14/2024
Create a comprehensive test suite for our project that supports PostgreSQL versions 13-17. This is a tough process, but it is worth it. After this, we can focus on the implementation of the required functions. The test suite will consist of two parts:

- GitHub Actions Integration: This part will automate checks for every pull request to ensure that any changes do not affect the normal execution of the project. We added a new workflow `ci.yml` in the `.github` directory. It will automatically install PostgreSQL versions 13-17 and test `pgmoneta_ext` in each version.

- Local Testing Setup: This part will handle the installation of all necessary dependencies and our extensions locally. It will then run tests on all existing functions to ensure they work correctly. The test functions will be implemented using [Check](https://libcheck.github.io/check/index.html), a unit testing framework for C. This local test also supports PostgreSQL versions 13-17. You can choose the version by passing a parameter when running `installation.c`.

For more details, please refer to the relevant [issues](https://github.com/pgmoneta/pgmoneta_ext/issues/9) and [pull requests](https://github.com/pgmoneta/pgmoneta_ext/pull/11).

### 06/06/2024
Having experience with `pg_switch_wal()`, I found that implementing the [CHECKPOINT](https://pgpedia.info/c/checkpoint-sql-command.html) function is much easier. `CHECKPOINT` is an SQL command that forces `PostgreSQL` to carry out an immediate [checkpoint](https://pgpedia.info/c/checkpoint.html), ensuring that the on-disk data files are up-to-date as of the command's execution.

We have separated this functionality by `PostgreSQL` version. For versions 13 and 14, this function can only be executed with `SUPERUSER` privileges. However, for version 15 and later, it can also be executed by the role `pg_checkpoint`.

For more details, you can check the [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/5) and [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/8).

### 06/03/2024
Finally, we have made a good start with our extension. I struggled a lot with the function `pg_switch_wal()`, which forces `PostgreSQL` to switch to a new [WAL](https://pgpedia.info/w/wal-write-ahead-logging.html). This function is crucial for our project `pgmoneta` to perform delta backups. It was really hard to set up this function. We wrapped it using our logic and described the security aspects.

The function behaves differently depending on the current role. If the role has `SUPERUSER` privileges, it will switch the `WAL` normally. However, if it doesn't have these privileges, it will do nothing, just return false and null. 

For more details, you can check the [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/4) and [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/7).

###  05/23/2024
We have created the functionality and documentation skeleton for the project. This will serve as the foundation. It includes:

1. Return the project version number as a `Datum` using a `PostgreSQL` extension. To achieve this functionality, four essential files have been added. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/1) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/2) for further details.
    - Control File: Specifies the extension’s name, version, and other metadata.
    - SQL File: Defines the `SQL` interface for the version retrieval function.
    - Code File: Implements the function in `C` to retrieve the project version number.
    - Makefile: Utilizes `CMake` to compile all the files and set up the extension.
2. Create basic documentation for users and developers. Users and developers have the option to generate a manual for this project, which offers a more readable format. Please refer to the code in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/6) and the associated [issue](https://github.com/pgmoneta/pgmoneta_ext/issues/3) for further details.

**Comment by Jubilee101** [comment]
Hi Chao, congrats on getting selected! In case you don't know, we've created the repository for pgmoneta_ext: https://github.com/pgmoneta/pgmoneta_ext. It's empty currently, so a lot could be worked on or ported from pgmoneta, including docs and ci flow. Perhaps a good start would be to just add a function to query the version of extension. That should lead you to have a basic structure of the extension to further work on. 

The ultimate goal for pgmoneta is to have peta-byte scale backup/restore. So when it comes to file streaming, it would be best if you could design for scalability, you probably already know that postgres (and pgmoneta as the client side) uses streaming replication protocal: https://www.postgresql.org/docs/current/protocol-replication.html. But, as a starter, streaming a file in the form of bytea is probably enough, baby steps :) 

We are dealing with harder and more open-ended problems now, so a lot could be discussed as we proceed. Glad to have you in the community!

**Reply by GuChad369** [reply]
I attempted to create a robust query function but was unsuccessful. 

Could you provide more details or the structure of the query function? Specifically, which file should I complete to implement this function? Starting from scratch has been challenging. 

I implemented a basic function to query the version of an extension, but it is too simplistic.

**Reply by jesperpedersen** [reply]
I would start with the skeleton of the extension, and a single function that return the version number in a `Datum`.

Small steps until you  get more familiar with extension development.

**Comment by jesperpedersen** [comment]
You can look at https://github.com/pgexporter/pgexporter_ext as a starting point...

**Comment by GuChad369** [comment]
I have established the skeleton of `pgmoneta_ext` and implemented the initial function that returns the version number in a `Datum`. Please review it in the [PR](https://github.com/pgmoneta/pgmoneta_ext/pull/1). 
However, there is a minor issue that I haven't been able to identify. I may need a little help with this.

**Comment by Jubilee101** [comment]
Hey, the extension seems to be on a good start. Remember that the current purpose of the extension is mostly delta backup.. and later the selective restore, which are most likely based on similar functions and infrastructures which you are currently building. 

You probably need a bigger picture on this, roughly I think some crucial problems are getting the WAL starting and ending point of a backup; building, receiving and comparing a manifest (some part of which I'm working on right now); and streaming the changed part to the client side. And on client side we also need new support to receive/restore/manage/retain delta backups. Remember that we intend to do this following postgres' protocals, so maybe a good reference would be how postgres does a base backup: https://www.interdb.jp/pg/pgsql10/01.html. 

I think it would be great if you have some rough milestones or questions you would want to work on or look into. Let us know if you want to create issues for problems worthwhile. Jesper and I are also looking into charting a detailed course. This is much more open ended than last year's gsoc, so I get that it could be quite dizzy in the beginning. Just let us know if you have any doubts or questions, or even don't know what to ask. Feel free to share your thoughts and ideas as well. We welcome discussions :).

**Reply by GuChad369** [reply]
From my understanding, our main task is to manipulate **WAL files** to achieve delta backup. The process involves:

1. Getting the **WAL** starting and ending points of a backup from the `primary`.
2. Building, receiving, and comparing a manifest.
3. Streaming the changed parts to the client side.
4. Using the three steps to perform the backup on the client side:
    - Issue the `pg_backup_start` command.
    - Take a snapshot of the database cluster.
    - Issue the `pg_backup_stop` command.
 
We have implemented our own `pg_switch_wal` and `CHECKPOINT`. Therefore, the following additional tasks are to create a **backup_label file** and a **backup history file**.

Is my understanding correct?

**Reply by jesperpedersen** [reply]
First, we need a way to test that the extension is working. Next, we will need a full manifest map of all the files on the pgmoneta side of things, and then a manifest map on the PostgreSQL side of things.

How we will get to delta backup if a question that we will explore later down the road

**Reply by Jubilee101** [reply]
This sounds about right. Note that Postgres doesn't really stop the world or taking a snapshot when making the backup, meaning there could be inconsistency inside -- a transaction may change some files that you have copied, so that change does not reflect on the copy, and others you have not yet copied, so the changes do reflect on them. But it doesn't matter since postgres replays WAL when starting on the backup. As per the [doc](https://www.postgresql.org/docs/current/continuous-archiving.html):

> We do not need a perfectly consistent file system backup as the starting point. Any internal inconsistency in the backup will be corrected by log replay (this is not significantly different from what happens during crash recovery). So we do not need a file system snapshot capability, just tar or a similar archiving tool.

Hence the need for `pg_backup_start` and `pg_backup_end`. which basically tells postgres where REDO should start and stop. Though one concern is the required privilege here -- both functions require superuser role, or being granted EXECUTE. So before you move forward, maybe a little investigation about why this is required, and if a user with only replication role is able to run the command -- I suspect so, since pg_basebackup triggers both functions internally. Otherwise we may need to look into other safer ways to get us a starting/ending point without operations that require too much previlege.

**Reply by Jubilee101** [reply]
And feel free to file issues about it. Getting the starting and ending point is quite crucial for the backup.

**Comment by jesperpedersen** [comment]
@GuChad369 Remember to update your weekly report

**Comment by GuChad369** [comment]
I have read the article and will state all my questions and discussions here. Below is my understanding of the new task. Is it correct? If there are any differences in understanding, please let me know. I will focus on how to solve this task based on this understanding.

`pg_dump`: This tool allows us to dump a single specified database. When using `pg_dump`, we can precisely control which database to back up without affecting others.

Replication Protocol: Typically used to create a base backup of the entire PostgreSQL instance, including all databases. This protocol does not allow for selective backups of individual databases, and our core project currently uses this method.

Current Issue: We need a way to identify the Object Identifiers (OIDs) of the specific databases we want to back up.

Task: We need to add a function to our extension that retrieves the OIDs of the desired databases. With this information, we can specify only the databases we want to back up. For example, if we only want to back up `mydb`, we can focus on its OID and exclude others.

**Reply by jesperpedersen** [reply]
We are using the `BASE_BACKUP` protocol, so we can't specify which database(s) to backup. But, if we had the OIDs then we could potentially remove the data, and update the system catalogues to get a running system.

So, we have to retrieve everything and then "repair" - or we have to look at a way where the extension can handle a backup of a single database and  send it back... For latter, it could maybe be done in the same way as `pg_dump` does it

**Reply by GuChad369** [reply]
I read the source code of [`pg_dump`](https://doxygen.postgresql.org/pg__dump_8c.html), and there are two main functions to implement the selective part:

- [`getTables`](https://doxygen.postgresql.org/pg__dump_8c.html#ae98b8dea58b417c8edce827f3326b013): This function read all the tables (no indexes) in the system catalogs.
- [`selectDumpableTable`](https://doxygen.postgresql.org/pg__dump_8c.html#a0ceaccc66d1af09ad40445a6e42b228f): This function decides whether we want to dump a specific table.

Although I haven't deeply read these functions, I understand that `pg_dump` first retrieves all tables and then, based on the command-line arguments provided by the user, selects specific tables using the `selectDumpableTable` function.

We can adopt a similar approach in our extension. First, retrieve all databases from the server and filter out the ones we want.

**Reply by jesperpedersen** [reply]
Yeah, in some way, because `BASE_BACKUP` sends a single TAR file with everything

**Reply by GuChad369** [reply]
Next, I will try to implement the function in the extension to return a specific OIDs.

**Reply by jesperpedersen** [reply]
Great, scan through the system catalogues and try and figure the queries that we need.

Maybe https://www.interdb.jp/pg/ can be of help

**Reply by GuChad369** [reply]
The exact queries we need will be implemented by other functions. I will first attempt to implement one function that returns the OID based on the database name. Then, we can try to find all queries related to this OID. This is my understanding.

**Comment by GuChad369** [comment]
I have researched the `pg_dump -Fc` path and tried to understand how to implement it. Now, I have a general idea of what it is and an outline of how it works.

The `-Fc` option specifies that the dump should be in a custom format. This custom format is a proprietary format that is more flexible and efficient. The custom format allows for the following features:

- **Compression**: The dump file is compressed, saving disk space.
- **Parallel Restore**: You can restore the dump in parallel, speeding up the restore process.
- **Selective Restoration**: You can restore individual tables or specific data from the dump, which isn't possible with plain SQL dumps.

### 1. **Relevant Files:**
   The key files related to `-Fc` are:
   - `pg_dump.c`: This is the main file where most of the logic for `pg_dump` resides.
   - `common.c` and `pg_backup_archiver.c`: These files contain functions related to the custom archive format.

### 2. **Examine the `-Fc` Option Handling:**
   The `-Fc` option is parsed in the `pg_dump.c` file in `main()` and `parseArchiveFormat()`.
- `main()`
  ```c
  ...
  case 'F':
    format = pg_strdup(optarg);
    break;
  
  ...
  /* Identify archive format to emit */
  archiveFormat = parseArchiveFormat(format, &archiveMode);
  ...
  ```
- `parseArchiveFormat()`
  ```c
  ...
  else if (pg_strcasecmp(format, "c") == 0)
    archiveFormat = archCustom;
  ...
  ```
### 3. **Follow the Code Path for Custom Format:**
   Once the format is set to `archCustom`, `pg_dump` will use the functions defined in `pg_backup_archiver.c` to handle the custom format. The key functions to look into are:
   - `CreateArchive()`: Create a new archive.
   - `WriteDataChunks()`: Write out all data (tables & LOs).
   - `WriteToc()`: Writes the table of contents (TOC) for the custom format.
   - `CloseArchive()`: Finalizes the archive.

The above functions that handle the custom format also involve serializing data and metadata into a binary format that `pg_restore` can later read and restore. 

### 4. **The `pg_backup_custom.c`:**
   The file `pg_backup_custom.c` contains the low-level implementation for the custom format.

### 5. **Summary:**
Currently, I've only learned up to this point and haven't figured out exactly how to implement these functions. I've been following the path to see which functions are involved. These are just the main processes, and there are also other files that help achieve this functionality.

In my opinion, if we try to simulate this functionality in our extension, it may take a significant amount of time to implement the entire process. 

For me, it feels almost like creating a completely new, large project.

**Reply by jesperpedersen** [reply]
Ok, it seem a bit too much work outside of our main goals with this GSoC project.

Ideally, it would be nice to backup a single database, but the protocol doesn't allow that, and doing all the work in the extension doesn't seem to be scalable right now. Maybe when we get deeper into the APIs a solution will appear, but for now lets just leave as "Investigated"
